Application/EdFi.Admin.DataAccess/Repositories/ClientAppRepo.cs

```diff
@@ -294,15 +294,107 @@ public void DeleteClient(string key)
             }
         }
 
-        public Application[] GetVendorApplications(int vendorId)
+        public IEnumerable<Vendor> GetVendors()
         {
             using (var context = _contextFactory.CreateContext())
             {
-                return context.Applications.Where(a => a.Vendor.VendorId == vendorId)
-                    .ToArray();
+                return context.Vendors.Include(u => u.VendorNamespacePrefixes).Include(a=>a.Applications)
+                    .Include(b=>b.Users)
+                    .ToList();
+            }
+        }
+
+        public IEnumerable<Application> GetApplications()
+        {
+            using (var context = _contextFactory.CreateContext())
+            {
+                return context.Applications.Include(u => u.ApplicationEducationOrganizations).Include(a=>a.Vendor).ToList();
+            }
+        }
+
+        public Vendor GetVendor(int vendorId)
+        {
+            using var context = _contextFactory.CreateContext();
+            return context.Vendors.FirstOrDefault(c => c.VendorId == vendorId);
+        }
+
+        public Application GetApplication(int applicationId)
+        {
+            using var context = _contextFactory.CreateContext();
+            var application = context.Applications
+                                .Where(x => x.ApplicationId == applicationId)
+                                .Include(x => x.ApplicationEducationOrganizations)
+                                .FirstOrDefault();
+            return application;
+        }
+
+        public void DeleteApplication(int applicationId)
+        {
+            using var context = _contextFactory.CreateContext();
+            var application = context.Applications.Include(a => a.ApiClients).First(x => x.ApplicationId == applicationId);
+
+            var applicationEducationOrganizations = context.ApplicationEducationOrganizations.Include(x => x.Application)
+                .Where(x => x.Application.ApplicationId == application.ApplicationId);
+
+            foreach (var applicationEducationOrganization in applicationEducationOrganizations)
+            {
+                context.ApplicationEducationOrganizations.Remove(applicationEducationOrganization);
+            }
+
+            foreach (var apiClient in application.ApiClients)
+            {
+                DeleteClient(apiClient.Key);
```

Comment:
Calling `DeleteClient` results in the error below because it's creating a nested Context (happens when you try to delete an Application that has a Sandbox associated to it).

Consider passing down the Context as an optional parameter (although I'm not a fan of that, can you think of an elegant solution?).

``
Microsoft.EntityFrameworkCore.DbUpdateConcurrencyException: 'The database operation was expected to affect 1 row(s), but actually affected 0 row(s); data may have been modified or deleted since entities were loaded. See http://go.microsoft.com/fwlink/?LinkId=527962 for information on understanding and handling optimistic concurrency exceptions.'
``

Reply:
I fixed it now